File: diagnostico.py
"""
diagnostico.py — Ferramenta de diagnóstico de seletores CSS por plataforma.

Port do diagnostico.js (v5 Node.js) para Python/Playwright.

Navega para cada plataforma, tira screenshot, salva o HTML completo
e analisa automaticamente quais seletores CSS retornam resultados —
essencial para atualizar seletores quando o site mudar o DOM.

Uso:
    python diagnostico.py                        → Todas as plataformas ativas
    python diagnostico.py --platform ml          → Só Mercado Livre
    python diagnostico.py --platform ml amazon   → ML + Amazon
    python diagnostico.py --visible              → Abre o browser visível
    python diagnostico.py --keyword "midea 12000 btus"

Saída em ./diagnostico/:
    {plataforma}_screenshot.png   → Screenshot da página carregada
    {plataforma}_page.html        → HTML completo para inspeção manual
    {plataforma}_analysis.json    → Seletores detectados + contagens
"""

import json
import sys
import time
import argparse
from pathlib import Path
from urllib.parse import quote_plus

from bs4 import BeautifulSoup
from loguru import logger
from playwright.sync_api import sync_playwright

from config import DIAGNOSTICO_DIR, USER_AGENTS

DEFAULT_KEYWORD = "ar condicionado 12000 btus inverter"

# ---------------------------------------------------------------------------
# Plataformas para diagnóstico
# ---------------------------------------------------------------------------
PLATFORMS = [
    {
        "id": "ml",
        "name": "Mercado Livre",
        "url_fn": lambda kw: f"https://lista.mercadolivre.com.br/{quote_plus(kw).replace('+','-').lower()}",
    },
    {
        "id": "amazon",
        "name": "Amazon Brasil",
        "url_fn": lambda kw: f"https://www.amazon.com.br/s?k={quote_plus(kw)}",
    },
    {
        "id": "magalu",
        "name": "Magazine Luiza",
        "url_fn": lambda kw: f"https://www.magazineluiza.com.br/busca/{quote_plus(kw)}/",
    },
    {
        "id": "shopee",
        "name": "Shopee",
        "url_fn": lambda kw: f"https://shopee.com.br/search?keyword={quote_plus(kw)}",
    },
    {
        "id": "leroy",
        "name": "Leroy Merlin",
        "url_fn": lambda kw: f"https://www.leroymerlin.com.br/busca?term={quote_plus(kw)}",
    },
]

# ---------------------------------------------------------------------------
# Seletores candidatos para teste automático
# ---------------------------------------------------------------------------
CANDIDATE_SELECTORS = [
    # Genéricos
    "article",
    ".product-card",
    ".product",
    ".search-result",
    "[data-testid*='product']",
    "[data-testid*='card']",
    # Mercado Livre — legado e Poly
    "li.ui-search-layout__item",
    ".ui-search-result",
    ".andes-card",
    ".poly-card",
    "ol.ui-search-layout li",
    # Amazon
    "div[data-component-type='s-search-result']",
    ".s-result-item",
    # Shopee
    ".shopee-search-item-result__item",
    "[data-sqe='item']",
    ".col-xs-2-4",
    # Magalu
    "[data-testid='product-card']",
    "li[data-testid]",
    "a[data-testid='product-card-container']",
    # Leroy Merlin / Fast Shop
    "[class*='ProductCard']",
    "[class*='product-card']",
    "[class*='productCard']",
]


def _build_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Diagnóstico de seletores CSS por plataforma",
        formatter_class=argparse.RawTextHelpFormatter,
    )
    parser.add_argument(
        "--platform", nargs="+",
        choices=[p["id"] for p in PLATFORMS],
        default=None,
        help="Plataformas a diagnosticar (padrão: todas)",
    )
    parser.add_argument(
        "--keyword", default=DEFAULT_KEYWORD,
        help=f"Keyword de teste (padrão: \"{DEFAULT_KEYWORD}\")",
    )
    parser.add_argument(
        "--visible", action="store_true", default=False,
        help="Exibir browser (útil para ver o que está sendo carregado)",
    )
    return parser.parse_args()


def _analyze_html(html: str) -> dict:
    """
    Analisa o HTML com BeautifulSoup e conta quantos elementos
    cada seletor candidato retorna.
    """
    soup = BeautifulSoup(html, "html.parser")
    container_counts = {}

    for sel in CANDIDATE_SELECTORS:
        try:
            count = len(soup.select(sel))
            if count > 0:
                container_counts[sel] = count
        except Exception:
            pass

    # Detecta elementos que contêm preço em formato brasileiro
    import re
    price_elements = []
    price_re = re.compile(r"R\$\s?[\d.]+[,.]?\d*")
    for el in soup.find_all(True):
        text = el.get_text(strip=True)
        if price_re.search(text) and len(text) < 30 and not el.find():
            classes = " ".join(el.get("class", []))[:80]
            entry = {"tag": el.name, "classes": classes, "sample": text[:30]}
            if entry not in price_elements:
                price_elements.append(entry)
        if len(price_elements) >= 10:
            break

    return {
        "container_candidates": container_counts,
        "price_elements": price_elements,
        "stats": {
            "total_elements": len(soup.find_all()),
            "links": len(soup.find_all("a")),
            "h2_count": len(soup.find_all("h2")),
            "title": soup.title.get_text() if soup.title else "",
        },
    }


def run_diagnostics(platforms: list, keyword: str, visible: bool) -> None:
    output_dir = Path(DIAGNOSTICO_DIR)
    output_dir.mkdir(parents=True, exist_ok=True)

    logger.info(f"\n{'═'*56}")
    logger.info("  DIAGNÓSTICO DE SELETORES")
    logger.info(f"{'═'*56}")
    logger.info(f"  Keyword: \"{keyword}\"")
    logger.info(f"  Browser visível: {'SIM' if visible else 'NÃO'}")
    logger.info(f"  Plataformas: {', '.join(p['name'] for p in platforms)}")
    logger.info(f"  Saída em: {output_dir.resolve()}")
    logger.info(f"{'═'*56}\n")

    with sync_playwright() as pw:
        browser = pw.chromium.launch(
            headless=not visible,
            args=["--no-sandbox", "--disable-setuid-sandbox",
                  "--disable-blink-features=AutomationControlled"],
        )

        for platform in platforms:
            url = platform["url_fn"](keyword)
            logger.info(f"\n▸ {platform['name']}")
            logger.info(f"  URL: {url}")

            context = browser.new_context(
                user_agent=USER_AGENTS[0],
                viewport={"width": 1366, "height": 768},
                locale="pt-BR",
                timezone_id="America/Sao_Paulo",
            )
            # Stealth básico
            context.add_init_script(
                "Object.defineProperty(navigator,'webdriver',{get:()=>undefined});"
            )
            page = context.new_page()

            try:
                page.goto(url, wait_until="domcontentloaded", timeout=60_000)
                try:
                    page.wait_for_load_state("networkidle", timeout=10_000)
                except Exception:
                    pass
                time.sleep(4)

                # Scroll para triggerar lazy-load
                for _ in range(5):
                    page.evaluate("window.scrollBy(0, 800)")
                    time.sleep(0.5)
                page.evaluate("window.scrollTo(0, 0)")
                time.sleep(2)

                # Screenshot
                screenshot_path = output_dir / f"{platform['id']}_screenshot.png"
                page.screenshot(path=str(screenshot_path), full_page=False)
                logger.info(f"  ✓ Screenshot: {screenshot_path}")

                # HTML completo
                html = page.content()
                html_path = output_dir / f"{platform['id']}_page.html"
                html_path.write_text(html, encoding="utf-8")
                logger.info(f"  ✓ HTML salvo: {html_path} ({len(html)//1024} KB)")

                # Análise de seletores
                analysis = _analyze_html(html)
                analysis_path = output_dir / f"{platform['id']}_analysis.json"
                analysis_path.write_text(
                    json.dumps(analysis, ensure_ascii=False, indent=2),
                    encoding="utf-8",
                )
                logger.info(f"  ✓ Análise: {analysis_path}")

                # Exibe seletores com >= 3 elementos (candidatos a container)
                found = sorted(
                    [(s, c) for s, c in analysis["container_candidates"].items() if c >= 3],
                    key=lambda x: -x[1],
                )
                if found:
                    logger.info("  ✓ SELETORES DE CONTAINER ENCONTRADOS:")
                    for sel, count in found[:6]:
                        logger.info(f"      {sel}  →  {count} elementos")
                else:
                    logger.warning(
                        "  ✗ Nenhum seletor padrão encontrado — "
                        "inspecione o HTML manualmente"
                    )

                # Exibe elementos de preço
                if analysis["price_elements"]:
                    logger.info("  ✓ ELEMENTOS DE PREÇO:")
                    for p in analysis["price_elements"][:3]:
                        logger.info(f"      <{p['tag']} class='{p['classes']}'>  →  \"{p['sample']}\"")

                logger.info(f"  ℹ Página: {analysis['stats']['title']}")

            except Exception as exc:
                logger.error(f"  ✗ ERRO: {exc}")
                try:
                    err_path = output_dir / f"{platform['id']}_error.png"
                    page.screenshot(path=str(err_path))
                    logger.info(f"  ✓ Screenshot de erro: {err_path}")
                except Exception:
                    pass
            finally:
                context.close()
                time.sleep(3)

        browser.close()

    logger.info(f"\n{'═'*56}")
    logger.info("  DIAGNÓSTICO COMPLETO")
    logger.info(f"{'═'*56}")
    logger.info(f"  Arquivos em: {output_dir.resolve()}")
    logger.info("")
    logger.info("  PRÓXIMOS PASSOS:")
    logger.info("  1. Abra os screenshots para verificar se a página carregou")
    logger.info("  2. Veja os _analysis.json para seletores detectados")
    logger.info("  3. Abra os _page.html no browser (F12) para inspecionar")
    logger.info("  4. Atualize _SELECTORS no scraper correspondente")
    logger.info("")
    logger.info("  DICA: Use --visible para ver o browser em tempo real")
    logger.info(f"{'═'*56}\n")


if __name__ == "__main__":
    args = _build_args()

    if args.platform:
        selected = [p for p in PLATFORMS if p["id"] in args.platform]
    else:
        selected = PLATFORMS

    run_diagnostics(
        platforms=selected,
        keyword=args.keyword,
        visible=args.visible,
    )

File: diagnostico.js
/**
 * ══════════════════════════════════════════════════════════════
 * DIAGNÓSTICO DE SELETORES — RAC Position Tracker
 * ══════════════════════════════════════════════════════════════
 * 
 * Este script navega para cada plataforma, tira screenshot e
 * salva o HTML da página para você inspecionar os seletores
 * corretos. Rode uma vez e abra os arquivos gerados na pasta
 * ./diagnostico/
 *
 * Uso:
 *   node diagnostico.js                  → Todas as plataformas
 *   node diagnostico.js --platform meli  → Só Mercado Livre
 *   node diagnostico.js --visible        → Abre o browser visível
 */

const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');

const KEYWORD = 'ar condicionado 12000 btus inverter';
const OUTPUT_DIR = './diagnostico';

const args = process.argv.slice(2);
const isVisible = args.includes('--visible');
const platformFilter = args.includes('--platform')
  ? args[args.indexOf('--platform') + 1]
  : null;

const PLATFORMS = [
  {
    id: 'meli',
    name: 'Mercado Livre',
    url: `https://lista.mercadolivre.com.br/${encodeURIComponent(KEYWORD).replace(/%20/g, '-')}`,
  },
  {
    id: 'amazon',
    name: 'Amazon Brasil',
    url: `https://www.amazon.com.br/s?k=${encodeURIComponent(KEYWORD)}`,
  },
  {
    id: 'shopee',
    name: 'Shopee',
    url: `https://shopee.com.br/search?keyword=${encodeURIComponent(KEYWORD)}`,
  },
  {
    id: 'magalu',
    name: 'Magazine Luiza',
    url: `https://www.magazineluiza.com.br/busca/${encodeURIComponent(KEYWORD)}/`,
  },
  {
    id: 'casasbahia',
    name: 'Casas Bahia',
    url: `https://www.casasbahia.com.br/${encodeURIComponent(KEYWORD)}/b`,
  },
  {
    id: 'google_shopping',
    name: 'Google Shopping',
    url: `https://www.google.com/search?tbm=shop&q=${encodeURIComponent(KEYWORD)}&gl=br&hl=pt-BR`,
  },
];

async function diagnose() {
  if (!fs.existsSync(OUTPUT_DIR)) fs.mkdirSync(OUTPUT_DIR, { recursive: true });

  let platforms = PLATFORMS;
  if (platformFilter) {
    platforms = platforms.filter(p => p.id.includes(platformFilter));
  }

  console.log('\n══════════════════════════════════════════════════');
  console.log('  DIAGNÓSTICO DE SELETORES');
  console.log('══════════════════════════════════════════════════');
  console.log(`  Keyword de teste: "${KEYWORD}"`);
  console.log(`  Browser visível: ${isVisible ? 'SIM' : 'NÃO'}`);
  console.log(`  Plataformas: ${platforms.map(p => p.name).join(', ')}`);
  console.log('══════════════════════════════════════════════════\n');

  const browser = await chromium.launch({
    headless: !isVisible,
    args: ['--no-sandbox', '--disable-setuid-sandbox'],
  });

  for (const platform of platforms) {
    console.log(`\n▸ ${platform.name}`);
    console.log(`  URL: ${platform.url}`);

    const context = await browser.newContext({
      userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36',
      viewport: { width: 1366, height: 768 },
      locale: 'pt-BR',
      timezoneId: 'America/Sao_Paulo',
    });
    const page = await context.newPage();

    try {
      // Navegar e esperar bastante para tudo carregar
      await page.goto(platform.url, { waitUntil: 'networkidle', timeout: 60000 });
      
      // Esperar extra para JS renderizar
      await page.waitForTimeout(5000);

      // Scroll para triggerar lazy loading
      await page.evaluate(async () => {
        for (let i = 0; i < 5; i++) {
          window.scrollBy(0, 800);
          await new Promise(r => setTimeout(r, 500));
        }
        window.scrollTo(0, 0);
      });
      await page.waitForTimeout(2000);

      // Screenshot
      const screenshotPath = path.join(OUTPUT_DIR, `${platform.id}_screenshot.png`);
      await page.screenshot({ path: screenshotPath, fullPage: false });
      console.log(`  ✓ Screenshot salvo: ${screenshotPath}`);

      // HTML completo
      const html = await page.content();
      const htmlPath = path.join(OUTPUT_DIR, `${platform.id}_page.html`);
      fs.writeFileSync(htmlPath, html, 'utf8');
      console.log(`  ✓ HTML salvo: ${htmlPath} (${(html.length / 1024).toFixed(0)} KB)`);

      // Análise automática de seletores candidatos
      const analysis = await page.evaluate(() => {
        const results = {};

        // Procurar containers de produto comuns
        const candidateSelectors = [
          // Genéricos
          '[data-testid*="product"]',
          '[data-testid*="card"]',
          '[data-component-type*="search-result"]',
          'article',
          '.product-card',
          '.product',
          '.search-result',
          // Mercado Livre
          '.ui-search-layout__item',
          '.ui-search-result',
          '.andes-card',
          'li.ui-search-layout__item',
          '.poly-card',
          'ol.ui-search-layout li',
          '.ui-search-layout--grid li',
          // Amazon
          'div[data-component-type="s-search-result"]',
          '.s-result-item',
          '.s-main-slot .s-result-item',
          // Shopee
          '.shopee-search-item-result__item',
          '[data-sqe="item"]',
          '.col-xs-2-4',
          // Magalu
          '[data-testid="product-card"]',
          'li[data-testid]',
          'a[data-testid="product-card-container"]',
          // Casas Bahia
          '.product-card',
          '[class*="ProductCard"]',
          '[class*="product-card"]',
          '[class*="productCard"]',
          'a[class*="product"]',
          // Google Shopping
          '.sh-dgr__gr-auto',
          '.sh-dlr__list-result',
          '.sh-pr__product-result',
          '[data-docid]',
        ];

        results.containerCandidates = {};
        for (const sel of candidateSelectors) {
          try {
            const count = document.querySelectorAll(sel).length;
            if (count > 0) {
              results.containerCandidates[sel] = count;
            }
          } catch (e) { /* seletor inválido */ }
        }

        // Procurar textos de preço
        const pricePattern = /R\$\s?[\d.]+[,.]?\d*/;
        const priceElements = [];
        document.querySelectorAll('*').forEach(el => {
          const text = el.textContent?.trim();
          if (text && pricePattern.test(text) && text.length < 30 && el.children.length === 0) {
            const classes = el.className?.toString().substring(0, 80) || '';
            const tag = el.tagName;
            if (!priceElements.some(p => p.classes === classes)) {
              priceElements.push({ tag, classes, sample: text.substring(0, 25) });
            }
          }
        });
        results.priceElements = priceElements.slice(0, 10);

        // Estatísticas gerais
        results.stats = {
          totalElements: document.querySelectorAll('*').length,
          links: document.querySelectorAll('a').length,
          images: document.querySelectorAll('img').length,
          h2Count: document.querySelectorAll('h2').length,
          h3Count: document.querySelectorAll('h3').length,
          title: document.title,
          url: window.location.href,
        };

        return results;
      });

      // Salvar análise
      const analysisPath = path.join(OUTPUT_DIR, `${platform.id}_analysis.json`);
      fs.writeFileSync(analysisPath, JSON.stringify(analysis, null, 2), 'utf8');
      console.log(`  ✓ Análise salvo: ${analysisPath}`);

      // Mostrar seletores encontrados
      const containers = analysis.containerCandidates;
      const found = Object.entries(containers).filter(([, count]) => count >= 3);
      if (found.length > 0) {
        console.log(`  ✓ SELETORES DE CONTAINER ENCONTRADOS:`);
        found.sort((a, b) => b[1] - a[1]);
        for (const [sel, count] of found.slice(0, 5)) {
          console.log(`      ${sel}  →  ${count} elementos`);
        }
      } else {
        console.log(`  ✗ Nenhum seletor padrão encontrado — verifique o HTML manualmente`);
      }

      // Mostrar preços encontrados
      if (analysis.priceElements.length > 0) {
        console.log(`  ✓ ELEMENTOS DE PREÇO ENCONTRADOS:`);
        for (const p of analysis.priceElements.slice(0, 3)) {
          console.log(`      <${p.tag} class="${p.classes}">  →  "${p.sample}"`);
        }
      }

      console.log(`  ℹ Página: ${analysis.stats.title}`);
      console.log(`  ℹ ${analysis.stats.totalElements} elementos, ${analysis.stats.links} links, ${analysis.stats.h2Count} h2`);

    } catch (err) {
      console.log(`  ✗ ERRO: ${err.message}`);

      // Tentar tirar screenshot mesmo com erro
      try {
        const errScreenshotPath = path.join(OUTPUT_DIR, `${platform.id}_error.png`);
        await page.screenshot({ path: errScreenshotPath });
        console.log(`  ✓ Screenshot de erro salvo: ${errScreenshotPath}`);
      } catch (_) {}
    }

    await context.close();

    // Delay entre plataformas
    await new Promise(r => setTimeout(r, 3000));
  }

  await browser.close();

  console.log('\n══════════════════════════════════════════════════');
  console.log('  DIAGNÓSTICO COMPLETO');
  console.log('══════════════════════════════════════════════════');
  console.log(`  Arquivos salvos em: ${path.resolve(OUTPUT_DIR)}`);
  console.log('');
  console.log('  PRÓXIMOS PASSOS:');
  console.log('  1. Abra os screenshots para verificar se a página carregou');
  console.log('  2. Abra os _analysis.json para ver seletores detectados');
  console.log('  3. Abra os _page.html no browser para inspecionar o HTML');
  console.log('  4. Atualize os seletores no config.js com os valores corretos');
  console.log('  5. Rode novamente: node diagnostico.js --platform XXXX');
  console.log('');
  console.log('  DICA: Use --visible para ver o browser em tempo real:');
  console.log('  node diagnostico.js --visible --platform meli');
  console.log('══════════════════════════════════════════════════\n');
}

diagnose().catch(err => {
  console.error('Erro fatal:', err.message);
  process.exit(1);
});
